main/Main.hs

```diff
@@ -50,42 +52,53 @@ options = Nixfmt
 format' :: Width -> FilePath -> Text -> Either String Text
 format' w path = first errorBundlePretty . format w path
 
-data Target = Target
-    { tDoRead :: IO Text
-    , tPath :: FilePath
-    , tDoWrite :: Text -> IO ()
-    }
+data Target
+    = StdioTarget
+    | FileTarget FilePath
+
+targetName :: Target -> String
+targetName StdioTarget = "<stdin>"
+targetName (FileTarget path) = path
+
+readTarget :: Target -> IO Text
+readTarget StdioTarget = TextIO.getContents
+readTarget (FileTarget path) = readFileUtf8 path
+
+filterTarget :: (Target -> IO Result) -> Target -> IO Result
+filterTarget operation StdioTarget = operation StdioTarget
+filterTarget operation (FileTarget path) = do
+    isSymlink <- try $ pathIsSymbolicLink path
```

Comment:
This will also catch async exceptions which you don’t want (this probably doesn’t matter in this specific case, but it’s better to place safe and future-proof). Use `try` for `safe-exceptions`.